models/tasks.js

```diff
@@ -88,24 +88,74 @@ const addDependency = async (data) => {
  *
  * @return {Promise<tasks|Array>}
  */
+
+const getBuiltTasks = async (tasksSnapshot) => {
+  const tasks = buildTasks(tasksSnapshot);
+  const promises = tasks.map(async (task) => fromFirestoreData(task));
+  const updatedTasks = await Promise.all(promises);
+  const taskPromises = updatedTasks.map(async (task) => {
+    task.status = TASK_STATUS[task.status.toUpperCase()] || task.status;
+    const taskId = task.id;
+    const dependencySnapshot = await dependencyModel.where("taskId", "==", taskId).get();
+    task.dependsOn = [];
+    dependencySnapshot.docs.forEach((doc) => {
+      const dependency = doc.get("dependsOn");
+      task.dependsOn.push(dependency);
+    });
+    return task;
+  });
+  const taskList = await Promise.all(taskPromises);
+  return taskList;
+};
+
+const fetchPaginatedTasks = async ({ status = "", size = TASK_SIZE, page, next, prev }) => {
+  try {
+    const initialQuery = status
+      ? tasksModel.where("status", "==", status).orderBy("title")
+      : tasksModel.orderBy("title");
+    let queryDoc = initialQuery;
+
+    if (prev) {
+      queryDoc = queryDoc.limitToLast(size);
+    } else {
+      queryDoc = queryDoc.limit(size);
+    }
+
+    if (page) {
+      const startAfter = size * page;
+      queryDoc = queryDoc.offset(startAfter);
+    } else if (next) {
+      const doc = await tasksModel.doc(next).get();
+      queryDoc = queryDoc.startAt(doc);
+    } else if (prev) {
+      const doc = await tasksModel.doc(prev).get();
+      queryDoc = queryDoc.endAt(doc);
+    }
+
+    const snapshot = await queryDoc.get();
+
+    const first = snapshot.docs[0];
+    const prevDoc = await initialQuery.endBefore(first).limitToLast(1).get();
```

Comment:
What does first have?
Does it not have a id?, do we need to fetch it in prevDoc

Reply:
The first has the reference to the first document in the range of documents to be returned.
For e.g. if the range of documents returned are 5th to 9th documents, then first will contain reference to the 5th document.
Using the reference to the 5th document (first among the list), we can fetch the previous document to it using `endBefore` which takes the 5th doc reference and the limitToLast which limits the documents from last to first.

Kindly do inform if I need to clarify more.